Spell checking in Constellation

Auriga2 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
Spell checking in now available in Constellation Data access view and Database importer. It can be added in other areas of Constellation as required.

In the right click context menu, `Turn On Spell Checking` check menu item can be used to toggle on/off spell checking of that field. It is set to `On` by default.

Incorrect spellings and some grammar errors are highlighted and underlined as the user types, except when any alpha numeric character or apostrophe is typed at the end. Left click on a highlighted word will pop up the suggestions list to choose from.

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs

<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?

<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
https://github.com/constellation-app/constellation/issues/870
<!-- Link any applicable issues here -->

Auriga2 commented:
- Unit tests are broken obviously, will fix them if we are happy with this change.
- Added a `CheckMenuItem` to toggle turn On/Off Spell Checking. Set it `On` by default. 
- Had to add a new `PasswordInputPane` to handle passwords separate to `StringParameterType`.
- Removed the redundant code for key pressed event handling as it was already inherited, in both `ValueInputPane` and `PasswordInputPane`.

TODO:

- Fix/add unit tests
- `JLanguageTool getSuggestedReplacements` populates a lot of irrelevant suggestions as well. Need to filter this.
- Custom css changes similar to blue background in `Test2` in `Test Parameters plugin` no longer works on the new text area. How shall we handle this?
- Add a `promptText`, although it's the same as the PluginParameter description displayed under the field name in DAV.
- Setting tool tip text-doesn't seem to work in the current build?

Delphinus8821 reviewed (changes requested):
Testing out the functionality and if you correct a spelling mistake the highlight and underline do not go away. The highlight and underline also come back if the first word written is spelt wrong and then the text is deleted. 

![spellcheckBug](https://github.com/constellation-app/constellation/assets/73083035/b2d91c86-e783-44af-877e-4a0ff94f72a3)

Auriga2 commented:
It's a limitation of the spell checker library. It seem to mark the first lowercase letter of a sentence before the actual spell checking. But hopefully, we might be able to implement a workaround along with filtering the long list of suggestions.

OrionsGuardian commented:
When you have some words already written in the text area, and you put the cursor between two existing words and start inserting a new word, it will underline/highlight the word before you've finished typing the word. 
It should wait until you've moved the cursor off of the word before checking the spelling of that word.

Auriga2 commented:
@OrionsGuardian I did that intentionally, `Incorrect spellings and some grammar errors are highlighted and underlined as the user types, except when any alpha numeric character or apostrophe is typed at the end.`

Because if you leave the text box without pressing any control character to trigger spell checking, the text could be marked incorrectly without the user knowing. This can happen most of the time the  user is editing in the middle of a word. 

Other option to fix is to trigger a final spell checking when loosing focus, but I thought it can still go unnoticed in a larger text.  Open for suggestions though.

antares1470 commented on CoreDependencies/src/ivy.xml:
```diff
@@ -81,6 +81,8 @@
         <dependency org="org.swinglabs" name="swingx" rev="1.6.1"/>
         <dependency org="org.fxmisc.richtext" name="richtextfx" rev="0.10.6"/>
+        <dependency org="org.languagetool" name="language-en" rev="5.7"/>
+        <dependency org="tech.units" name="indriya" rev="1.3" force="true" />
```
I note from the project.xml that we already had v2.0.2 of this. Was there a reason why this specific version was needed?

Auriga2 replied:
Yes, remember it took long to get the LanguageTool libworking in Consty, it was causing issues with v2 of indriya . Arcturus2 identified the problem. pom.xml mentions 1.3 of indriya but our ivy was getting v2. Also the force="true" was also needed.

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/ValueInputPane.java:
```diff
@@ -105,92 +99,89 @@ public ValueInputPane(final PluginParameter<?> parameter, final int defaultWidth
 
+
+            Platform.runLater(() -> {
+                Text t = (Text) field.lookup(".text");
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+    public void checkSpelling() {
+        textArea.clearStyles();
+        misspells.clear();
+        if (!turnOffSpellChecking && !StringUtils.isBlank(textArea.getText())) {
```
`StringUtils.isNotBlank` for that 2nd boolean

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+        if (!turnOffSpellChecking && !StringUtils.isBlank(textArea.getText())) {
+            try {
+                matches = langTool.check(textArea.getText());
+            } catch (IOException ex) {
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+            try {
+                matches = langTool.check(textArea.getText());
+            } catch (IOException ex) {
+                Logger.getLogger(SpellChecker.class.getName()).log(Level.SEVERE, null, ex);
```
can we make the logger a constant for the class (similar to what has been done on other classes in the code. Also if the message could instead be `ex.getLocalisedMessage()` rather than null

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+            matches.forEach(match -> {
+                final int start = match.getFromPos();
+                final int end = match.getToPos();
+                String misspell = textArea.getText().substring(start, end);
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+            return false;
+        }
+
+        for (RuleMatch match : matches) {
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        for (RuleMatch match : matches) {
+            int start = match.getFromPos();
+            int end = match.getToPos();
```
both final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,158 @@
+        return false;
+    }
+
+    public void turnOffSpellChecking(boolean turnOffSpellChecking) {
```
parameter final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+        });
+
+        // Set the right click context menu
+        ContextMenu contextMenu = addRightClickContextMenu();
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+     * underline and highlight the text from start to end.
+     */
+    public void highlightText(final int start, final int end) {
+        String underlineAndHighlight = "-rtfx-background-color:derive(yellow,-30%);"
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+        //TODO
+    }
+
+//    public void bindAutoCompletion(List<String> suggestions) {
+//        this.setSpellCheckCallback(word -> {
+//            if (!suggestions.isEmpty()) {
+//                return suggestions.stream()
+//                        .filter(suggestion -> suggestion.startsWith(word))
+//                        .toArray(String[]::new);
+//            }
+//            return null;
+//        });
+//    }
```
remove if not intending to use

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+     * the middle of a sentence.
+     */
+    private boolean canCheckSpelling(final String newText) {
+        return newText.length() > 0 && !(this.getCaretPosition() == this.getText().length()
```
can you make the first condition `!nextText.isEmpty()`

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    private ContextMenu addRightClickContextMenu() {
+        ContextMenu contextMenu = new ContextMenu();
+
+        MenuItem undoMenuItem = new MenuItem("Undo");
+        MenuItem redoMenuItem = new MenuItem("Redo");
+        MenuItem cutMenuItem = new MenuItem("Cut");
+        MenuItem copyMenuItem = new MenuItem("Copy");
+        MenuItem pasteMenuItem = new MenuItem("Paste");
+        MenuItem deleteMenuItem = new MenuItem("Delete");
+        MenuItem selectAllMenuItem = new MenuItem("Select All");
```
these can all be final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+        selectAllMenuItem.setOnAction(e -> this.selectAll());
+
+        // CheckMenuItem to toggle turn On/Off Spell Checking. On by default
+        CheckMenuItem toggleSpellCheckMenuItem = new CheckMenuItem("Turn On Spell Checking");
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    public void autoComplete(final List<String> suggestions) {
+        Popup popup = new Popup();
+        popup.setWidth(this.getWidth());
+        ListView<String> listView = new ListView<>();
```
both can be final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+            popup.getContent().clear();
+            listView.getItems().clear();
+
+            if (!StringUtils.isBlank(input) && !suggestions.isEmpty()) {
```
`StringUtils.isNotBlank`

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,243 @@
+                });
+
+                // Show the popup under this text area
+                if (listView.getItems().size() > 0) {
```
`!listView.getItems().isEmpty`

antares1470 reviewed (changes requested):
Haven't been able to test yet but here are some feedback code-wise

antares1470 reviewed (changes requested):
Have finally been able to give it a test. I'll add a few things to go with the other observations already made:
1) It seems the initial spellcheck seems to freeze the application momentarily. Not sure if that is an issue with the library, or the way it is being used (or just the system I was testing on)
2) I actually spelt the first word correctly and after it did a spellcheck it came back as wrong because I didn't include a capital letter (seems to presume I'm writing a sentence and hence the first word should be capitalised). (This one may be unavoidable but is an observation)
3) If you fix the spelling of a word, it is corrected for all instances of the word (which may not necessarily be desired).
![spellcheckIssue](https://github.com/constellation-app/constellation/assets/62311243/2abff0e1-5032-4ee1-be81-03fb79056829)

Regarding @OrionsGuardian's comment, I think I'm inclined to agree. I did a test with Word and what I noticed was that it would remove a squiggly line as soon as a valid word was written, but it doesn't generate a squiggly line (and all the stuff that comes with spellcheck fixing) until you moved away from the word (e.g. create whitespace character, step over whitespace character). I think that behaviour is possibly more desirable than the immediate feedback of your word being wrong before you've even finished.

Auriga2 commented:
@antares1470 
1. Yes this is a limitation with the library. The very first initializing is slow but after that, it is fast. 
2. Yes this is also a limitation with the library. I'm checking whether we can implement a workaround.
3. Bug, fixed now.

Auriga2 commented:
1. Fixed now by calling a spell check on a random text when initializing. Sort of a dodgy way but had to fix this somehow, now that the spell checking is only triggered after typing the first word
2. Easy fix. Just had to disable the rule.
3. Fixed
4. Similar to Microsoft Word, words no longer mark until typing is finished, confirmed by any non alpha numeric character except apostrophe symbol ('). It removes highlight when a highlighted word is corrected by typing, but if a correct word is edited it wouldn't trigger a spell checking until moved away from the word (Slightly different to word: e.g. create whitespace character at the end of word, step over whitespace character at the end or move to the start of the word with the left arrow key). 

5. Added a constellation preference (Enabled by default) and a pluginParameter  option (Disabled by default and should be enabled in the code per plugin parameter) to enable/disable spell checking.
If both of these are set, the right click context menu option `Turn On Spell Checking` will be available and it can be used to turn off spell checking dynamically. That could be handy if a text contains words from other languages etc. making the highlighting annoying.

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/PasswordInputPane.java:
```diff
@@ -0,0 +1,124 @@
+    private final TextInputControl field;
+    private final boolean required;
+    private int comboBoxWidth = EMPTY_WIDTH;
+    private static final Logger LOGGER = Logger.getLogger(ValueInputPane.class.getName());
```
Should be `PasswordInputPane` rather than `ValueInputPane`

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,196 @@
+     * selected row.
+     */
+    private void initialize() {
+        for (Rule rule : langTool.getAllRules()) {
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,196 @@
+     */
+    private void initialize() {
+        for (Rule rule : langTool.getAllRules()) {
+            if (rule.getId().equals("UPPERCASE_SENTENCE_START")) {
+                langTool.disableRule(rule.getId());
```
assuming the rule id is unique, could be worth adding a break after this since presumably every other rule will return false at the if condition

Auriga2 replied:
Using the same loop to keep track of spellingCheckRule now. Hence cannot break now. But wondering whether to use 2 loops with break instead.?

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,196 @@
+        popup = new Popup();
+        suggestions.getSelectionModel().selectedItemProperty().addListener((obs, oldValue, newValue) -> {
+            if (newValue != null) {
+                StringBuilder builder = new StringBuilder(textArea.getText());
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,196 @@
+        //initialize langtool to prevent the spell checking being too slow at the first word after loading costy
+        try {
+            matches = langTool.check("random text");
+        } catch (final IOException ex) {
+            LOGGER.log(Level.SEVERE, ex.getLocalizedMessage());
+        }
+
+        //initialize popup
+        popup = new Popup();
```
perhaps the rest of the code after the catch should be moved into the try block given we're logging an error in the catch at which point we probably don't want to execute anything further in the function

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellChecker.java:
```diff
@@ -0,0 +1,196 @@
+        if (!turnOffSpellChecking && StringUtils.isNotBlank(textArea.getText())) {
+            try {
+                matches = langTool.check(textArea.getText());
+            } catch (final IOException ex) {
+                LOGGER.log(Level.SEVERE, ex.getLocalizedMessage());
+            }
+
+            matches.forEach(match -> {
```
same here

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,240 @@
+    /**
+     * underline and highlight the text from start to end.
+     */
+    public void highlightText(final int start, final int end) {
+
+
```
remove the extra whitespace here

antares1470 reviewed (changes requested):
Looking good in testing. A couple of other things though:
1) Are you able to add a whatsnew for this?
2) Are you able to add some unit tests for some of the classes you added?

OrionsGuardian commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/text/SpellCheckingTextArea.java:
```diff
@@ -0,0 +1,240 @@
+        this.setStyle(0, this.getText().length(), clearStyle);
+    }
+
+    public boolean isWordUnderCursorHighlighted(int index) {
```
`final`

OrionsGuardian commented:
![image](https://github.com/constellation-app/constellation/assets/104604488/05876da3-bcc6-43be-9af2-7d7317bcd1ff)

Having Done some testing (see above image), there are a few questions/suggestions for the spellchecking functionality.
Can we provide some configuration options or additional control to the user?
* Enable/Disable grammar checking
* Enable/Disable word background colour highlighting (and allow user to choose a highlight colour)
* Allow user to add words to a personal dictionary, so they are no longer highlighted. 
* Show/popup the words in the personal dictionary via the options screen, allowing words to be removed.

antares1470 commented:
> Having Done some testing (see above image), there are a few questions/suggestions for the spellchecking functionality. Can we provide some configuration options or additional control to the user?
> 
>     * Enable/Disable grammar checking
> 
>     * Enable/Disable word background colour highlighting (and allow user to choose a highlight colour)
> 
>     * Allow user to add words to a personal dictionary, so they are no longer highlighted.
> 
>     * Show/popup the words in the personal dictionary via the options screen, allowing words to be removed.

I think those ideas may be good as evolutions to the initial solution (so possibly a ticket or 2 for after this is merged in)

antares1470 commented on CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt:
```diff
@@ -1,6 +1,12 @@
+== 2024-01-31 Spell checking in Constellation
+<p>Spell checking in now available in text areas of Constellation Data access view (Extract Types and Extract Words Plugins) and Database importer. It can be added in other text areas as requested.
+When enabled, incorrect spellings and some grammar errors are highlighted and underlined. Left Clicking on a highlighted word/phrase will popup a suggestions list to choose from. An "Ignore All" button is available for spelling errors which will ignore the same error in the current text area.</p>
+<p>Constellation Option  "Enable spell checking in specified text areas" (Enabled by default) and the plugin Parameter option "setSpellCheckEnabled"(Disabled by default and should be enabled in the code per plugin parameter) are used to enable/disable spell checking.
```
I would omit the plugin parameter option note since users won't actually see this. Instead maybe just refer fields where it has been allowed

antares1470 reviewed (commented):
I'm not exactly sure what is going on, but I noted when I tried to run Constellation with this branch built that the load time was considerably slower (only had Data Access View open along with Welcome page). I ended up doing some time comparisons with master (hit run, waited for Constellation to load completely, then closed immediately after and noted the time period in the console) and confirmed that on my system, the load time difference was in excess of 40 secs on average.

Not sure what the exact cause is but I think we probably want to try and reduce that if we can.

Auriga2 commented:
Initializing another LanguageTool in the static block fixed the problem. It improved the speed but not as much as I wanted.

antares1470 reviewed (changes requested):
The load time is looking a bit better comparatively. Thanks @Auriga2

Just having done some testing, I noted that the Ignore All feature isn't quite working as intended. If I misspell a word, choose to ignore all, and then use that same misspelling later, it tries to correct in a way that seems wrong (see GIF). Compared this to MS Word, which when doing the same thing, doesn't add any additional squiggles
![spellcheckIssue2](https://github.com/constellation-app/constellation/assets/62311243/e57499f5-d4aa-45a0-a3fb-0141cec78d12)

antares1470 reviewed (changes requested):
I'm happy to leave repeated phrase detection given I couldn't think of many instances where you would want the repetition. That being said, I've found another bug. If you already corrected a misspelled word, the next time you use the same misspelled word, it will automatically use your previous correction as soon as you try to address it, irrespective of how you actually choose to correct in that instance.
![spellcheckIssue3](https://github.com/constellation-app/constellation/assets/62311243/384da531-fb07-4878-86d6-4541c080ebea)

OrionsGuardian commented:
![image](https://github.com/constellation-app/constellation/assets/104604488/2be74560-a4a5-427b-9c95-fada198e0c7f)

The specified text areas should either be listed here, or we should provide a button/link which will give a popup showing which text areas have spell checking enabled.

antares1470 reviewed (commented):
It's looking good in testing. Just noting there are some merge conflicts that will need to be resolved

Delphinus8821 reviewed (changes requested):
Looks like its working really well. 

Just a couple of things:
There is a huge amount of blank space between the spell checking preference and the colorblind one. 
![image](https://github.com/constellation-app/constellation/assets/73083035/122c4eae-d965-4388-b4b5-66ecc5a8a9a1)

I think it could be useful to have a note on the spellcheck preference to say you have to left click the word for the spellcheck options to appear. I know this is written in the what's new text but users may miss this, I was confused as I expected the spellcheck options to be in a right click menu.

It would also be good to have something on the application options help page about the spellcheck functionality - can probably use the same text as the what's new entry.

Also looks like there might be some unit tests failing.

Auriga2 commented:
@Delphinus8821 This is the error I get, all in `constellation.utilities.geospatial`, but don't think it's related to my changes.
![image](https://github.com/constellation-app/constellation/assets/64826928/b968b228-f148-4d23-a9bf-eb732bed7a55)

antares1470 reviewed (commented):
![spellcheckIssue4](https://github.com/constellation-app/constellation/assets/62311243/0a19ea7e-a4bf-4c86-97e6-41f9de41808e)

See above GIF. Without knowing for sure, it seems that if there aren't any spelling suggestions then there is no pop up that comes up. Wondering whether we switch this to instead popup that there are no spelling suggestions? My initial thought was that it wasn't working or wasn't registering my click so having some sort of feedback for the user would be useful

antares1470 commented:
> @Delphinus8821 This is the error I get, all in `constellation.utilities.geospatial`, but don't think it's related to my changes. ![image](https://private-user-images.githubusercontent.com/64826928/311073577-b968b228-f148-4d23-a9bf-eb732bed7a55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4NzA2MzgsIm5iZiI6MTcwOTg3MDMzOCwicGF0aCI6Ii82NDgyNjkyOC8zMTEwNzM1NzctYjk2OGIyMjgtZjE0OC00ZDIzLWE5YmYtZWI3MzJiZWQ3YTU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA4VDAzNTg1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhYzQ0NmRhNjM2OWU0OWNkNzUxODg0MGRlZDUwZTgzZjY3NDk3MWNjY2ViMTcxMjE2ODBmNmE2OWI5ZmEyYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7WZIu_LwCG-fTadzlESaRlTewEDYYgKEA4wo0TEd7-I)

Those tests have been known to fail occasionally (for some reason) and usually rerunning the job gets past that. That being said, I tried rerunning several times and it always failed in the same spot.

To figure out what was going on, I tried running the tests locally in netbeans (on this branch) and discovered that tests failed to run altogether (process fork failure). I switched over to master and running the tests there is fine. I don't think the issue is necessarily those shape unit tests, but it does seem something is causing issues with the unit tests

antares1470 reviewed (commented):
The spellcheck still seems to be working with the update. @Auriga2 do you happen to know where else indriya was being used (the other version)? Might want to check that the other functionality still works as well

Auriga2 commented:
@antares1470 Not sure of places other than constellation.utilities.geospatial where the unit tests failed.

antares1470 reviewed (approved):
Subject to merge conflicts being resolved, I think this is good to go
